### PR TITLE
Implement YAML validation and status code validation

### LINKED ISSUE #41: YAML validation
- Official *YAML Schema* doesn't exist
  - We'll do ad-hoc validation

### PR DESCRIPTION
**Pull request by kmizu**
- YAML validation
  - missing key
  - type of pattern is valid
  - value of pattern is valid
  - unknown key
    - except "id", "pattern", "message", "justification", "tests"

 - Status code
  - CONFIG_FILE_NOT_FOUND: 3
  - CONFIG_FILE_SYNTAX_ERROR: 4
  - CONFIG_FILE_UNKNOWN_ERROR: 5

Close #41

### CODE REVIEW

**Comment by kmizu** [comment]
@ybiquitous could you review it ?

**Comment by ybiquitous** [comment]
Sorry for the delay. 🙇  I'll review this later..

**Comment by ybiquitous** [comment]
I'd like to describe the exit codes, but do you have any good ideas where to be described?
Markdown or `--help`?

**Comment by ybiquitous** [comment]
Also, I will configure Sider to reduce too many issues. If you have any ideas, please give me feedback about Checkstyle or PMD settings.

**Comment by kmizu** [comment]
> I'd like to describe the exit codes, but do you have any good ideas where to be described?
> Markdown or `--help`?

OK.  I'll describe it in `--help`.

**Comment by kmizu** [comment]
https://github.com/sider/JavaSee/pull/102#issuecomment-689306995

It has been done in https://github.com/sider/JavaSee/pull/102/commits/1911735bc156d272b9efc8175a97ee44ed0ef37a

**Comment by ybiquitous** [comment]
I've added `sider.yml` via 1a03d3c3f196b3abba0afc37b3cef5939e753b83. Could you please merge the `master` branch to this feature branch?

**Comment by ybiquitous** [comment]
When an empty `javasee.yml` is given, a bit weird message is output:

```console
$ javasee test
YAML file has unknown errornull
```

**Comment by kmizu** [comment]
> When an empty `javasee.yml` is given, a bit weird message is output:
> 
> ```
> $ javasee test
> YAML file has unknown errornull
> ```

It seems that the weird message was caused by `NullPointerException`.  I'll fix it.

**Comment by ybiquitous** [comment]
Thanks. In addition, even if a user writes `rule` instead of `rules` (typo), `javasee test` succeeds:

```console
$ cat javasee.yml
rule:
  - id: foo
    pattern: bar
    message: baz

$ javasee test
Checking rule id uniqueness...
Checking rule patterns...
Tested 0 rules with 0 tests
  All tests green!
```

I think it is not a bit user-friendly, so can JavaSee output an error message instead of success?

**Comment by kmizu** [comment]
@ybiquitous 
> Thanks. In addition, even if a user writes `rule` instead of `rules` (typo), `javasee test` succeeds:
> 
> ```
> $ cat javasee.yml
> rule:
>   - id: foo
>     pattern: bar
>     message: baz
> 
> $ javasee test
> Checking rule id uniqueness...
> Checking rule patterns...
> Tested 0 rules with 0 tests
>   All tests green!
> ```
> 
> I think it is not a bit user-friendly, so can JavaSee output an error message instead of success?

This problem also has been fixed in https://github.com/sider/JavaSee/pull/102/commits/3a2c23eee42a87a60cd06004edc24ce629311d63 .  Now,  the following message is shown:

```
 ~/work/JavaSee (git yaml-validation) 
[k.mizushima]$ cat javasee.yml 
rule:
        - id: foo
          pattern: bar
          message: baz

 ~/work/JavaSee (git yaml-validation) 
[k.mizushima]$ java -jar build/libs/JavaSee-all.jar check
Fatal error: com.github.sider.javasee.Exceptions$MissingKeyException: Missing key: rules
Backtrace:
  com.github.sider.javasee.Config$Factory.config(Config.java:34)
  com.github.sider.javasee.Config.load(Config.java:21)
  com.github.sider.javasee.command.CheckCommand.start(CheckCommand.java:64)
  com.github.sider.javasee.Main.lambda$main$2(Main.java:66)
  java.base/java.util.Optional.map(Optional.java:265)
  com.github.sider.javasee.Main.main(Main.java:66)
```

**Comment by ybiquitous** [comment]
Thanks. I could confirm it was fixed! 👍 

```console
$ cat javasee.yml
rules:
  - id: foo
    pattern: bar
    message: baz
$ javasee test
YAML ERROR:
Missing key: rules
```

**Comment by ybiquitous** [comment]
Oh, should the exit code be `4` (`CONFIG_FILE_SYNTAX_ERROR`) instead of `1` (`ERROR`)?

```console
$ cat javasee.yml
rule:
  - id: aaa
    pattern: bar
    message: baz

$ javasee test
YAML ERROR:
Missing key: rules

$ echo $?
1
```

**Comment by kmizu** [comment]
> Oh, should the exit code be `4` (`CONFIG_FILE_SYNTAX_ERROR`) instead of `1` (`ERROR`)?
> 
> ```
> $ cat javasee.yml
> rule:
>   - id: aaa
>     pattern: bar
>     message: baz
> 
> $ javasee test
> YAML ERROR:
> Missing key: rules
> 
> $ echo $?
> 1
> ```

Currently, `CONFIG_FILE_SYNTAX_ERROR` means that it is syntactic invalid YAML file.  In this case, the YAML file doesn't follow schema.  Should I redefine `CONFIG_FILE_SYNTAX_ERROR` ?

**Comment by ybiquitous** [comment]
I think it natural to handle both a broken YAML syntax and an invalid YAML schema as `CONFIG_FILE_SYNTAX_ERROR`. Do you have any good ideas?

**Comment by kmizu** [comment]
@ybiquitous 
> I think it natural to handle both a broken YAML syntax and an invalid YAML schema as `CONFIG_FILE_SYNTAX_ERROR`. Do you have any good ideas?

In the commit https://github.com/sider/JavaSee/pull/102/commits/50013c929f9f805dc953f10a74b0984a900dec75, `CONFIG_FILE_SCHEMA_ERRROR` is added to handle such case. See the following:

image

**Comment by ybiquitous** [comment]
Make sense. Thank you. 👍

**Comment by ybiquitous** [comment]
I could confirm this change works expectedly, thanks.

I've closed trivial issues, but there are 21 open issues still:
https://sider.review/gh/repos/171470727/pulls/102?statuses%5B%5D=opened

So, could you please fix or close them? If needed, please leave a comment when you close an issue.

File: src/main/java/com/github/sider/javasee/lib/Libs.java [hunk 1]
```diff
@@ -1,9 +1,16 @@
 package com.github.sider.javasee.lib;
 
+import com.github.sider.javasee.Exceptions;
+
 public class Libs {
     public static interface SupplierWithException<T> {
         T get() throws Exception;
     }
+
+    public static interface SupplierWithPolymorphicException <T, U extends Exception> {
```
Comment (kmizu) [inline src/main/java/com/github/sider/javasee/lib/Libs.java:10]: > __[Checkstyle]__ '<' is preceded with whitespace. ([view](https://sider.review/gh/repos/171470727/pulls/102#issue-4548921))
>
> <details>
>
> | Rule | Severity |
> | ---- | -------- |
> | `GenericWhitespaceCheck` | `error` |
>
> References:
> - <https://checkstyle.org/config_whitespace.html#GenericWhitespace>
>
> You can [close this issue](https://sider.review/gh/repos/171470727/pulls/102#issue-4548921) if no need to fix it. [Learn more](https://help.sider.review/getting-started/working-with-issues).
>
> </details>

Such

**Review by ybiquitous — APPROVED** [review]
Thank you so much for your awesome work! 👍

**Pull request merged**